### PR TITLE
Revert "Revert "[core] Nested task support via task depth + backpressure" (#20438)"

### PR DESCRIPTION
**Pull request by wuisawesome**
## Why are these changes needed?

This PR reverts the previous revert with the following minor changes.

1. Worker capping is off by default.
2. The cap feature flag is on the for the tests that explicitely require it. 

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/tests/test_worker_capping.py [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+
+    # Assert that last_wwait / second_last ~= 2, with a healthy buffer since ci
+    # is noisy.
+    assert 1 < last_wait / second_last < 4
+    assert 2 < last_wait < 4
+
+
+if __name__ == "__main__":
+    import pytest
+    os.environ["RAY_worker_cap_initial_backoff_delay_ms"] = "1000"
```
Comment (wuisawesome) [inline python/ray/tests/test_worker_capping.py:184]: This class explicitly tests this PR's behavior.

File: src/ray/common/ray_config_def.h [hunk 2]
```diff
@@ -106,18 +106,24 @@ RAY_CONFIG(int64_t, max_lineage_bytes, 1024 * 1024 * 1024)
...
 RAY_CONFIG(bool, preallocate_plasma_memory, false)
 
+/// We place a soft cap on the number of tasks of a given scheduling class that
+/// can run at once to limit the total nubmer of worker processes. After the
+/// specified interval, the new task above that cap is allowed to run. The time
+/// before the next tasks (above the cap) are allowed to run increases
+/// exponentially. The soft cap is needed to prevent deadlock in the case where
+/// a task begins to execute and tries to `ray.get` another task of the same
+/// class. Setting the delay to 0 disables this feature.
+RAY_CONFIG(int64_t, worker_cap_initial_backoff_delay_ms, 0)
```
Comment (wuisawesome) [inline src/ray/common/ray_config_def.h:116]: Off by default now.

File: src/ray/raylet/scheduling/cluster_task_manager_test.cc [hunk 3]
```diff
@@ -185,6 +185,24 @@ class MockTaskDependencyManager : public TaskDependencyManagerInterface {
   std::unordered_set<TaskID> blocked_tasks;
 };
 
+class FeatureFlagEnvironment : public ::testing::Environment {
```
Comment (wuisawesome) [inline src/ray/raylet/scheduling/cluster_task_manager_test.cc:188]: Keep the flag on for its unit tests.

**Comment by ericl** [comment]
Tests are failing.

**Comment by iycheng** [comment]
Do we need to cherry-pick the fix?

**Review by fishbone — APPROVED** [review]

**Comment by iycheng** [comment]
Do we have a plan to roll it out by default?

**Comment by ericl** [comment]
Are we sure tests aren't still flaky?

**Comment by wuisawesome** [comment]
> Are we sure tests aren't still flaky?

I'm not really sure how to prove that they aren't flaky, but the feature is off by default in this PR, so I'd hope it is? 

> Do we have a plan to roll it out by default?

I think we want to merge this off by default, then merge #20449 to turn it on (maybe after running the microbenchmarks and  confirming no regression there). 

We can also wait a few days to make sure this doesn't break anything in the nightly tests while it's off by default.

**Comment by ericl** [comment]
How about running the unit tests a couple more times?

**Comment by wuisawesome** [comment]
Sure, this round looks good, will rerun all of them.

**Comment by wuisawesome** [comment]
Looks like we should be good to remerge this.

**Pull request merged**